core/src/main/java/org/elasticsearch/cockroach/CockroachTransportService.java

```diff
@@ -0,0 +1,404 @@
+        }
+    }
+
+    /**
+     * Asks the master node to assign a response to the running cockroach task
+     * <p>
+     * This method is used to start a cockroach task.
+     *
+     * @param task     the caller task
+     * @param uuid     the uuid of the running task
+     * @param response response that should be assigned to the task
+     * @param listener the listener that will be called when the task is updated
+     */
+    public <Response extends CockroachResponse> void sendResponse(final Task task, final String uuid, final Response response,
+                                                                  final ActionListener<Empty> listener) {
+        sendResponse(task, uuid, response, null, listener);
+    }
+
+    /**
+     * Asks the master node to assign a failure to the running cockroach task
+     * <p>
+     * This method is used to to indicate that the executor node successfully finished the running task.
+     *
+     * @param task     the caller task
+     * @param uuid     the uuid of the running task
+     * @param failure  failure that should be assigned to the task
+     * @param listener the listener that will be called when the task is updated
+     */
+    public void sendFailure(final Task task, final String uuid, final Exception failure, final ActionListener<Empty> listener) {
+        sendResponse(task, uuid, null, failure, listener);
+    }
+
+    private <Response extends CockroachResponse> void sendResponse(final Task task, final String uuid, final Response response,
+                                                                   final Exception failure, final ActionListener<Empty> listener) {
+        try {
+            String nodeId = clusterService.state().getNodes().getLocalNodeId();
+            if (clusterService.state().nodes().isLocalNodeElectedMaster()) {
+                cockroachActionCoordinator.finishCockroachTask(uuid, response, failure, listener);
+            } else {
+                UpdateCockroachTaskRequest<Response> startRequest = new UpdateCockroachTaskRequest<>(uuid, response, failure);
+                startRequest.setParentTask(nodeId, task.getId());
+                transportService.sendRequest(clusterService.state().nodes().getMasterNode(), UPDATE_COCKROACH_ACTION_NAME, startRequest,
+                    new ActionListenerEmptyTransportResponseHandler(ThreadPool.Names.GENERIC, listener));
+            }
+        } catch (Exception t) {
+            listener.onFailure(t);
+        }
+    }
+
+    /**
+     * Asks the master node to remove the running cockroach task
+     * <p>
+     * This method is called when the caller node has acknowledged the receive the response.
+     *
+     * @param task     the caller task
+     * @param uuid     the uuid of the running task
+     * @param listener the listener that will be called when the task is removed
+     */
+    public void acknowledgeResponse(final Task task, final String uuid, final ActionListener<Empty> listener) {
```

Comment:
I think a better name would be `acknowledgeResponseReceipt` or something similar